### DISCUSSION TITLE
Using EmitCompilerGeneratedFiles and CompilerGeneratedFilesOutputPath to keep track of generated files on source control

### DISCUSSION THREAD

**Discussion by MulletPr00f** [dotnet/roslyn#75141] [Q&A] [answered]
I'm using this configuration in my *.csproj:

```
<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
<CompilerGeneratedFilesOutputPath>GEN_PATH</CompilerGeneratedFilesOutputPath>
```

I have a few questions:

1) Is there a way to emit generated files only to one specific generator referenced in the project? I don't want to do that to _all_ generators, definitely not to the generators I'm not responsible for, such as the default generators referenced by the Microsoft SDKs.

2) Is there a way to set `CompilerGeneratedFilesOutputPath` to a relative path of the input file? Examples:

- `\Directory_A\File_A.ext` would generate the file `\Directory_A\File_A.generated.ext.cs`
- `\Directory_B\File_B.ext` would generate the file `\Directory_B\File_B.generated.ext.cs`

**Comment by 333fred** [comment, marked as answer]
1. No, we haven't found any need for this. Can you elaborate on why you care to output some generators, but not all?
2. No.

**Reply by MulletPr00f** [reply]
I shouldn't be responsible for handling generated files other than mine. 

When I hook up a code generation library in my project which does not emit compiler generated files, they have to be generated everytime they are opened in workspace solution, right? That means if we use automated builds with Azure Pipelines, they will be generated on every build, resulting on longer pipeline build runs than regular ones. 

In the other hand, if didn't have to generate those files everytime they are opened in a solution, and have them on source control, we could disable the project reference to the code generation library with some condition like this:

`<ProjectReference Include="..\MyGenerators\MyGenerators.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" Condition="'$(Configuration)'=='Debug'" />`

Doing so, our automated builds would take less time. I'm talking about 100+ microservices projects.

**Reply by 333fred** [reply]
It sounds like there's a bit of confusion here. `EmitCompilerGeneratedFiles` does not control whether generators are run; there's no material impact to your build times from this switch unless your backing storage is _particularly_ slow (like, I'd have to actually see the benchmark that shows even a 5400 RPM spinning rust disk causing a real delay here to believe that it's a problem). What you're talking about, conditionally enabling generators in DEBUG, is likely going to cause much more pain for you than it will possibly save in build time.

**Reply by MulletPr00f** [reply]
Sorry, I think you misundestood me. You said "EmitCompilerGeneratedFiles does not control whether generators are run.". Yes, absolutely. But if I enable it and keep the generated files on source control, I can disable the code generation on release configuration.

Let's say I have a code generation library called `MyGenerators` that generates source code files using as input *.razor files. I'm using it on a client project called `MyProject`, which has only 27 *.razor files at the moment, but will probably grow to hundreds of *.razor files. Related to `MyProject` there are some scenarios:

**Scenario A**
ProjectReference: no reference to `MyGenerators`.
Build config: Debug.
Build result:
```
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 2:02 AM and took 03.195 seconds ==========
```
After build: no generated files.

**Scenario B**
ProjectReference: reference to `MyGenerators` without any condition.
EmitCompilerGeneratedFiles: false.
Build config: Debug (Release times are very similar)
Build result:
```
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 2:05 AM and took 11.423 seconds ==========
```
After build: 27 generated files.

**Scenario C**
ProjectReference: reference to `MyGenerators` without any condition.
EmitCompilerGeneratedFiles: true
Build config: Debug
Build result:
```
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 2:13 AM and took 11.711 seconds ==========
```
After build: 27 generated files that can (and will) be added to source control by some developer.

**Scenario D (Release - automated build)**
ProjectReference: reference to `MyGenerators` with condition `$(Configuration)'=='Debug'`.
EmitCompilerGeneratedFiles: true
Build config: **Release**
Build result:
```
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 2:16 AM and took 04.237 seconds ==========
```
After build: no generated files by the build itself. But I already have 27 generated files that were previously added to source control.

You can notice an increase in the build time when `MyGenerators` is referenced. It's not much for a single project in one developer's machine, with a small number of input files, but my concern is Scenario B with Release configuration on an automated build, with hundreds of input files to be generated.

The problem is how the CI pipeline of automated build works: it creates a temporary solution with the modified projects (and if it's references are added as project references, those projects also). So it will always generate those source code files, cause the files are always new to the workspace. Incremental generation won't matter on this scenario, only on a dev's machine. Regarding the pipeline itself, it can have a lot of integration processes, like unit testing, health checks, so it will take longer than just a build to run. Let's say a small project will take something around 3-5 minutes to finish a pipeline. This pipeline get's triggered on every developer's check-in in the source control, and only starts when the previous has finished. So having these generated files not being generated on a release build, but added to the source control and built like regular *.cs files, would mitigate this decrease of performance.

So, having said that and considering this time penalty, I think Roslyn source code generators as it is maybe aren't the best choice for our environment.

A workaround could be not calling `context.AddSource` on my source output action, and just disabling RS1035 and creating a *.cs file by hand on a desired path of my choice, and add that file to source control, idk.

**Reply by 333fred** [reply]
I'd say that the first step is investigating which generators are taking up that time. You can use https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Report%20Analyzer%20Format.md to add analyzer timing information to a binlog; you can then use that to start looking into which generators are taking up time. IMO, attempting to have generators automatically run in DEBUG mode, but then _not_ in RELEASE, is just setting you up for a world of pain and unintentional footguns of "well, it worked on my machine", because someone forgot to do a complete DEBUG build locally and check in updated generated files.

**Reply by 333fred** [reply]
Oh, and absolutely _do not_ disable RS1035 and do IO in your generator. This will break your build's up-to-date checking, as well as your VS experience. The compiler has discussed ways to actually _block_ IO in generators and analyzers, and we may be much harder on it in the future.

**Reply by MulletPr00f** [reply]
> You can use https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Report%20Analyzer%20Format.md to add analyzer timing information to a binlog; you can then use that to start looking into which generators are taking up time.

I'll try that. Even though there's only one generator in this project reference and it does nothing special, it's generating an empty class.

>IMO, attempting to have generators automatically run in DEBUG mode, but then _not_ in RELEASE, is just setting you up for a world of pain and unintentional footguns of "well, it worked on my machine", because someone forgot to do a complete DEBUG build locally and check in updated generated files.

I get your point, but that shouldn't really happen in this case, as it would break the build if a generated file doesn't exist. It would be a partial class that's generated and supposed to have implementation used in *.razor or *.razor.cs files.

**Reply by jasonmalinowski** [reply]
> it would break the build if a generated file doesn't exist. It would be a partial class that's generated and supposed to have implementation used in *.razor or *.razor.cs files.

The scary bit wouldn't be a missing file, but a file that is stale in some way or another in a way that compiles (since the methods are fine) but the contents are off. Then you end up code that works fine locally but fails in production or whatever.

**Reply by jkant** [reply]
I have a use case where I'd like one generator that creates C#/.NET code and does not emitted generated files, and another generator that emits HLSL code which is not part of the .NET solution and should be thought of more as content files - but is controlled by the same logic as the logic that generates the C# code. Is there a recommended way to approach this type of workflow @333fred ?

**Reply by 333fred** [reply]
No, there is no recommended way to approach that workflow in a Roslyn Source Generator. SGs generate either C# or VB code, not code for alternative languages.

**Reply by jkant** [reply]
Fair enough! Thanks for the reply.
